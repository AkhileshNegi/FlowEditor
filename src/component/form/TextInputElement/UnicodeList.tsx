import * as React from 'react';
import * as uniqid from 'uniqid';
import { UnicodeCharMap, pluralize } from './helpers';
import * as styles from './UnicodeList.scss';

interface UnicodeListProps {
    unicodeChars: UnicodeCharMap;
}

export const utfWarning = (num: number) =>
    ` Note that your message may require more segments to send because it contains the following ${pluralize(
        num,
        'character'
    )}:`;

const UnicodeList: React.SFC<UnicodeListProps> = ({ unicodeChars }) => {
    const chars = Object.keys(unicodeChars);
    if (chars.length) {
        return (
            <React.Fragment>
                {utfWarning(chars.length)}
                <div className={styles.unicodeList}>
                    {chars.map((char, idx) => (
                        <span key={uniqid()} className={styles.unicodeChar}>
                            {char}
                        </span>
                    ))}
                </div>
            </React.Fragment>
        );
    }
    return null;
};

export default UnicodeList;
