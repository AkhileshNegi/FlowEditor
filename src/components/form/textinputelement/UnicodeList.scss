.unicode-char {
    display: inline-block;
    width: 25px;
    height: 25px;
    text-align: center;
    background-color: #b3dffe;
    border-radius: 4px;
    margin-top: 3px;
    margin-right: 3px;
    line-height: 25px;
    vertical-align: top;
    overflow: hidden;
    border-right: 1px solid #fff;
    border-bottom: 1px solid #fff;
}

.unicode-list {
    display: flex;
    flex-wrap: wrap;
    overflow-y: auto;
    max-height: 70px;
    border: 1px solid #eee;
    border-radius: 4px;
    padding: 5px;
    margin: 5px 0;
    width: 203px;
}
