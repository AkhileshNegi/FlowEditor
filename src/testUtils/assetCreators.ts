import { languageToAsset } from '~/components/flow/actions/updatecontact/helpers';
import { Methods } from '~/components/flow/routers/webhook/helpers';
import { Operators } from '~/config/operatorConfigs';
import { Types } from '~/config/typeConfigs';
import {
    AnyAction,
    BroadcastMsg,
    CallWebhook,
    Case,
    ChangeGroups,
    Contact,
    Exit,
    Field,
    Flow,
    FlowNode,
    Group,
    Label,
    Router,
    RouterTypes,
    SendEmail,
    SendMsg,
    SetContactChannel,
    SetContactField,
    SetContactLanguage,
    SetContactProperty,
    SetRunResult,
    StartFlow,
    StartFlowArgs,
    StartFlowExitNames,
    StartSession,
    SwitchRouter,
    UINode,
    Wait,
    WaitTypes
} from '~/flowTypes';
import { AssetType } from '~/services/AssetService';
import { RenderNode } from '~/store/flowContext';
import { capitalize } from '~/utils';

const { assets: groupsResults } = require('~/test/assets/groups.json');
const languagesResults = require('~/test/assets/languages.json');

/**
 * Create a select control option
 */
export const createSelectOption = ({ label }: { label: string }) => ({
    label: capitalize(label.trim()),
    labelKey: 'name',
    valueKey: 'id'
});

export const createSendMsgAction = ({
    uuid = 'send_msg-0',
    text = 'Hey!',
    all_urns = false
}: {
    uuid?: string;
    text?: string;
    // tslint:disable-next-line:variable-name
    all_urns?: boolean;
} = {}): SendMsg => ({
    type: Types.send_msg,
    uuid,
    text,
    all_urns
});

export const createSendEmailAction = ({
    uuid = 'send_email-0',
    subject = 'New Sign Up',
    body = '@run.results.name just signed up.',
    addresses = ['jane@example.com']
}: {
    uuid?: string;
    subject?: string;
    body?: string;
    addresses?: string[];
} = {}): SendEmail => ({
    uuid,
    type: Types.send_email,
    subject,
    body,
    addresses
});

export const createCallWebhookAction = ({
    uuid = 'call_webhook-0',
    url = 'https://www.example.com',
    method = Methods.GET
}: {
    uuid?: string;
    url?: string;
    method?: Methods;
} = {}): CallWebhook => ({
    uuid,
    type: Types.call_webhook,
    url,
    method
});

export const createStartSessionAction = ({
    uuid = 'start-session-0',
    groups = [{ uuid: 'group-0', name: 'Cat Fanciers' }, { uuid: 'group-1', name: 'Cat Facts' }],
    contacts = [
        { uuid: 'contact-0', name: 'Kellan Alexander' },
        { uuid: 'contact-1', name: 'Norbert Kwizera' },
        { uuid: 'contact-2', name: 'Rowan Seymour' }
    ],
    flow = {
        uuid: 'flow_uuid',
        name: 'Flow to Start'
    }
}: {
    uuid?: string;
    groups?: Group[];
    contacts?: Contact[];
    flow?: Flow;
} = {}): StartSession => ({
    uuid,
    groups,
    contacts,
    flow,
    type: Types.start_session
});

export const createBroadcastMsgAction = ({
    uuid = 'send_broadcast-0',
    groups = [{ uuid: 'group-0', name: 'Cat Fanciers' }, { uuid: 'group-1', name: 'Cat Facts' }],
    contacts = [
        { uuid: 'contact-0', name: 'Kellan Alexander' },
        { uuid: 'contact-1', name: 'Norbert Kwizera' },
        { uuid: 'contact-2', name: 'Rowan Seymour' }
    ],
    text = 'Hello World'
}: {
    uuid?: string;
    groups?: Group[];
    contacts?: Contact[];
    text?: string;
} = {}): BroadcastMsg => ({
    uuid,
    groups,
    contacts,
    text,
    type: Types.send_broadcast
});

export const createAddGroupsAction = ({
    uuid = 'add_contact_groups-0',
    groups = groupsResults
}: { uuid?: string; groups?: Group[] } = {}): ChangeGroups => ({
    uuid,
    type: Types.add_contact_groups,
    groups
});

export const createStartFlowAction = ({
    uuid = 'start_flow-0',
    flow = {
        name: 'Colors',
        uuid: 'colors-0'
    }
}: {
    uuid?: string;
    flow?: {
        name: string;
        uuid: string;
    };
} = {}): StartFlow => ({
    type: Types.start_flow,
    uuid: 'start-flow-0',
    flow: {
        name: capitalize(flow.name.trim()),
        uuid
    }
});

export const createSetContactNameAction = ({
    uuid = 'set_contact_name-0',
    name = 'Jane Goodall'
}: {
    uuid?: string;
    name?: string;
} = {}): SetContactProperty => ({
    uuid,
    name,
    type: Types.set_contact_name
});

export const createSetContactFieldAction = ({
    uuid = 'set_contact_field-0',
    field = {
        key: 'age',
        name: 'Age'
    },
    value = '25'
}: {
    uuid?: string;
    field?: Field;
    value?: string;
} = {}): SetContactField => ({
    uuid,
    field,
    value,
    type: Types.set_contact_field
});

export const createSetContactLanguageAction = ({
    uuid = 'set_contact_language-0',
    language = 'eng'
}: {
    uuid?: string;
    language?: string;
} = {}): SetContactLanguage => ({
    uuid,
    language,
    type: Types.set_contact_language
});

export const createSetContactChannelAction = ({
    uuid = 'set_contact_channel-0',
    channelName = 'Twilio Channel'
}: {
    uuid?: string;
    channelName?: string;
} = {}): SetContactChannel => ({
    uuid,
    channel: {
        uuid,
        name: channelName
    },
    type: Types.set_contact_channel
});

export const createSetRunResultAction = ({
    uuid = 'set_run_result-0',
    name = 'Name',
    value = 'Grace',
    category = ''
}: {
    uuid?: string;
    name?: string;
    value?: string;
    category?: string;
} = {}): SetRunResult => ({
    uuid,
    name,
    value,
    category,
    type: Types.set_run_result
});

// tslint:disable-next-line:variable-name
export const createCase = ({
    uuid,
    type,
    exit_uuid,
    args = []
}: {
    uuid: string;
    type: Operators;
    exit_uuid: string;
    args?: string[];
}) => ({
    uuid,
    type,
    exit_uuid,
    arguments: args
});

export const createExit = ({
    uuid = 'exit-0',
    name = null,
    destination_node_uuid = 'node-1'
}: {
    uuid?: string;
    name?: string;
    // tslint:disable-next-line:variable-name
    destination_node_uuid?: string;
} = {}) => ({
    uuid,
    name,
    destination_node_uuid
});

// tslint:disable-next-line:variable-name
export const createWait = ({ type, timeout }: { type: WaitTypes; timeout?: number }) => ({
    type,
    ...(timeout ? { timeout } : {})
});

// tslint:disable-next-line:variable-name
export const createRouter = (result_name?: string): Router => ({
    type: RouterTypes.switch,
    ...(result_name ? { result_name } : {})
});

export const createSwitchRouter = ({
    cases,
    operand = '@input',
    default_exit_uuid = null
}: {
    cases: Case[];
    operand?: string;
    // tslint:disable-next-line:variable-name
    default_exit_uuid?: string;
}) => ({
    ...createRouter(),
    cases,
    operand,
    default_exit_uuid
});

export const createRenderNode = ({
    actions,
    exits,
    uuid = 'node-0',
    router = null,
    wait = null,
    ui = {
        position: { left: 0, top: 0 }
    }
}: {
    actions: AnyAction[];
    exits: Exit[];
    uuid?: string;
    router?: Router | SwitchRouter;
    wait?: Wait;
    ui?: UINode;
}): RenderNode => ({
    node: {
        actions,
        exits,
        uuid,
        ...(router ? { router } : ({} as any)),
        ...(wait ? { wait } : ({} as any))
    },
    ui,
    inboundConnections: null
});

export const createFlowNode = ({
    actions,
    exits,
    uuid = 'node-0',
    router = null,
    wait = null
}: {
    actions: AnyAction[];
    exits: Exit[];
    uuid?: string;
    router?: Router | SwitchRouter;
    wait?: Wait;
}): FlowNode => ({
    actions,
    exits,
    uuid,
    ...(router ? { router } : ({} as any)),
    ...(wait ? { wait } : ({} as any))
});

export const createWaitRouterNode = ({
    exits,
    cases,
    uuid = 'wait-router',
    timeout
}: {
    exits: Exit[];
    cases: Case[];
    timeout?: number;
    uuid?: string;
}): RenderNode =>
    createRenderNode({
        actions: [],
        exits,
        uuid,
        router: createSwitchRouter({
            cases
        }),
        wait: createWait({ type: WaitTypes.msg, timeout })
    });

export const createStartFlowNode = (
    startFlowAction: StartFlow,
    uuid: string = 'start_flow_node-0',
    // tslint:disable-next-line:variable-name
    flow_uuid?: string,
    exitUUIDs: string[] = ['exit1', 'exit2']
): RenderNode =>
    createRenderNode({
        actions: [startFlowAction],
        exits: [
            createExit({
                uuid: exitUUIDs[0],
                name: StartFlowExitNames.Complete,
                destination_node_uuid: 'destination-completed'
            }),
            createExit({
                uuid: exitUUIDs[1],
                name: StartFlowExitNames.Expired,
                destination_node_uuid: 'destination-expired'
            })
        ],
        uuid,
        router: createSwitchRouter({
            cases: [
                createCase({
                    uuid: 'start_flow_case-0',
                    type: Operators.has_run_status,
                    exit_uuid: exitUUIDs[0],
                    args: [StartFlowArgs.Complete]
                }),
                createCase({
                    uuid: 'start_flow_case-1',
                    type: Operators.has_run_status,
                    exit_uuid: exitUUIDs[1],
                    args: [StartFlowArgs.Expired]
                })
            ],
            operand: '@child',
            default_exit_uuid: null
        }),
        wait: createWait({ type: WaitTypes.flow })
    });

export const createGroupsRouterNode = (
    groups: Group[] = groupsResults,
    uuid: string = 'split_by_groups-0'
): RenderNode =>
    createRenderNode({
        actions: [],
        exits: groups.map((group, idx) =>
            createExit({ uuid: group.uuid, name: group.name, destination_node_uuid: `node-${idx}` })
        ),
        uuid,
        router: createSwitchRouter({
            cases: groups.map((group, idx) =>
                createCase({
                    uuid: `split_by_group-${idx}`,
                    type: Operators.has_group,
                    exit_uuid: group.uuid,
                    args: [group.uuid]
                })
            ),
            operand: '@contact.groups',
            default_exit_uuid: null
        }),
        wait: createWait({ type: WaitTypes.group })
    });

export const getGroupOptions = (groups: Group[] = groupsResults) =>
    groups.map(({ name, uuid }) => ({
        name,
        id: uuid
    }));

export const getGroups = (sliceAt: number, groups: Group[] = groupsResults) =>
    groups
        .map(({ name, uuid }) => ({
            name,
            id: uuid
        }))
        .slice(sliceAt);

export const createAddLabelsAction = (labels: Label[]) => ({
    type: Types.add_input_labels,
    uuid: `labels-action-uuid-${labels.length}`,
    labels
});

export const English = { name: 'English', id: 'eng', type: AssetType.Language };

export const Spanish = { name: 'Spanish', id: 'spa', type: AssetType.Language };

export const SubscribersGroup = {
    name: 'Subscriber',
    id: 'subscribers_group',
    type: AssetType.Group
};

export const FeedbackLabel = { name: 'Feedback', id: 'feedback_label', type: AssetType.Label };

export const languages = languagesResults.assets.map((language: any) => languageToAsset(language));
