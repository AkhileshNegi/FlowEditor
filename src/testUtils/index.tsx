// TODO: Remove use of Function
// tslint:disable:ban-types
import { mount, ReactWrapper, shallow, ShallowWrapper } from 'enzyme';
import { object } from 'prop-types';
import * as React from 'react';
import * as config from '../../__test__/config';
import { ConfigProviderContext, endpointsPT, flowPT, languagesPT } from '../config';
import { FlowDefinition, FlowEditorConfig } from '../flowTypes';
import { AppState, createStore, initialState } from '../store';
import { getBaseLanguage } from '../utils';

export interface Resp {
    results: Array<{ [key: string]: any }>;
    total?: number;
}

export interface QueryString {
    [key: string]: string;
}

const initialContext: ConfigProviderContext = {
    endpoints: (config as FlowEditorConfig).endpoints,
    languages: (config as FlowEditorConfig).languages,
    flow: (config as FlowEditorConfig).flow
};

const initialTestState: AppState = {
    ...initialState,
    ...{
        flowContext: {
            ...initialState.flowContext,
            definition: require('../../__test__/flows/colors.json') as FlowDefinition
        }
    },
    ...{
        flowEditor: {
            ...initialState.flowEditor,
            editorUI: {
                ...initialState.flowEditor.editorUI,
                language: getBaseLanguage((config as FlowEditorConfig).languages)
            }
        }
    }
};

/**
 * Compose setup method for component tests
 */
export const createSetup = <P extends {}>(
    Component: React.ComponentClass | React.SFC,
    baseProps: P = {} as any,
    baseDuxState: AppState | Partial<AppState> = initialState,
    baseContext: ConfigProviderContext = initialContext,
    connectedChildren: boolean = false
) => (
    shallowRender: boolean = true,
    propOverrides: P | Partial<P> = {},
    duxStateOverrides = {} as AppState,
    contextOverrides = {} as ConfigProviderContext | Partial<ConfigProviderContext>,
    childContextTypeOverrides: { [key: string]: Function } = {}
) => {
    // Waiting on https://github.com/Microsoft/TypeScript/pull/1328
    const props = Object.assign({}, baseProps, propOverrides);
    const context = connectedChildren
        ? Object.assign({}, baseContext, {
              store: createStore(Object.assign({}, baseDuxState, duxStateOverrides))
          })
        : baseContext;
    // tslint:disable-next-line:ban-types
    const wrapper = (shallowRender ? (shallow as Function) : (mount as Function))(
        <Component {...props} />,
        {
            context,
            childContextTypes: Object.assign(
                {},
                {
                    store: object,
                    endpoints: endpointsPT,
                    languages: languagesPT,
                    flow: flowPT
                },
                childContextTypeOverrides
            )
        }
    );

    return {
        wrapper,
        props,
        context,
        instance: wrapper.instance()
    };
};
export const createSpy = (obj: Object | React.ComponentClass) => (instanceMethod: string) =>
    jest.spyOn((obj as React.ComponentClass).prototype || obj, instanceMethod as any);

/**
 * Wait for promises in queue to resolve
 */
export const flushPromises = () => new Promise(resolve => setImmediate(resolve));

/**
 * Restore spy mocks (distinct from mocks created w/ jest.fn()).
 * Use to declaratively restore multiple spy mocks,
 * otherwise just call .mockRestore() on lone spy.
 */
export const restoreSpies = (...spies: jest.SpyInstance[]) =>
    spies.forEach(spy => spy.mockRestore());

/**
 * NOTE: borrowed from EventBrite: https://github.com/eventbrite/javascript/blob/master/react/testing.md#finding-nodes
 * Finds all instances of components in the rendered `componentWrapper` that are DOM components
 * with the `data-spec` attribute matching `name`.
 * @param {ReactWrapper|ShallowWrapper} componentWrapper - Rendered componentWrapper (result of mount, shallow, or render)
 * @param  {string} snacName - Name of `data-spec` attribute value to find
 * @param {string|Function} (Optional) typeFilter - (Optional) Expected type of the wrappers (defaults to all HTML tags)
 * @returns {ReactWrapper|ReactWrapper[]|ShallowWrapper|ShallowWrapper[]} Matching DOM components
 */
export const getSpecWrapper = (
    componentWrapper: ReactWrapper<{}, {}> | ShallowWrapper<{}, {}>,
    specName: string,
    attributeName: string = 'data-spec'
): any => {
    return componentWrapper.find(`[${attributeName}="${specName}"]`);
};

const DEBUG = false;

export const timeStart = (name: string) => {
    if (DEBUG) {
        console.time(name);
    }
};

export const timeEnd = (name: string) => {
    if (DEBUG) {
        console.timeEnd(name);
    }
};
