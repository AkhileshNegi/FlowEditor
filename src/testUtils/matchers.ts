import { RenderNode } from '../store/flowContext';
import { dump } from '../utils';
import { Exit } from '../flowTypes';
import { Store } from 'redux';
import { AppState } from '../store';

interface MatchResult {
    message: () => string;
    pass: boolean;
}

function toHaveInboundConnections<T>(this: jest.MatcherUtils, received: RenderNode): MatchResult {
    if (Object.keys(received.inboundConnections).length > 1) {
        return {
            message: () => `Node ${received.node.uuid} has an inbound connection`,
            pass: true
        };
    }

    return {
        message: () => `Node${received.node.uuid} has no inbound connections`,
        pass: false
    };
}

function toHaveInboundFrom<T>(
    this: jest.MatcherUtils,
    received: RenderNode,
    expected: Exit
): MatchResult {
    for (const exitUUID of Object.keys(received.inboundConnections)) {
        if (exitUUID === expected.uuid) {
            return {
                message: () =>
                    `Node ${received.node.uuid} incorrectly has inbound from ${expected.uuid}`,
                pass: true
            };
        }
    }

    return {
        message: () =>
            `Node ${received.node.uuid} doesn't have inbound connection from ${expected.uuid}`,
        pass: false
    };
}

function toPointTo<T>(this: jest.MatcherUtils, received: Exit, expected: RenderNode): MatchResult {
    if (received.destination_node_uuid === expected.node.uuid) {
        return {
            message: () => `${received.uuid} incorrectly points to ${expected.node.uuid}`,
            pass: true
        };
    }

    return {
        message: () => `Expected ${received.uuid} to point to ${expected.node.uuid}`,
        pass: false
    };
}

function toHaveExitWithDestination<T>(this: jest.MatcherUtils, received: RenderNode): MatchResult {
    for (const exit of received.node.exits) {
        if (exit.destination_node_uuid !== null) {
            return {
                message: () => `Exit ${exit.uuid} in node ${received.node.uuid} has a destination`,
                pass: true
            };
        }
    }

    return {
        message: () => `All exits for ${received.node.uuid} are disconnected`,
        pass: false
    };
}

function toHaveExitThatPointsTo<T>(
    this: jest.MatcherUtils,
    received: RenderNode,
    expected: RenderNode
): MatchResult {
    for (const exit of received.node.exits) {
        if (exit.destination_node_uuid === expected.node.uuid) {
            return {
                message: () =>
                    `${exit.uuid} incorrectly has exit that points to ${expected.node.uuid}`,
                pass: true
            };
        }
    }

    return {
        message: () =>
            `Expected ${received.node.uuid} to have an exit that points to ${expected.node.uuid}`,
        pass: false
    };
}

function toHavePayload<T>(
    this: jest.MatcherUtils,
    store: any,
    actionType: string,
    expectedPayload: any
): MatchResult {
    const payload = JSON.stringify(expectedPayload);
    for (const actionTaken of store.getActions()) {
        if (actionTaken.type === actionType) {
            if (JSON.stringify(actionTaken.payload) === payload) {
                return {
                    message: () =>
                        `Result contained action type ${actionType} with payload ${payload}`,
                    pass: true
                };
            }
        }
    }

    return {
        message: () => `Could not find action type ${actionType} with payload ${payload}`,
        pass: false
    };
}

function toHaveReduxAction<T>(
    this: jest.MatcherUtils,
    store: any,
    actionType: string
): MatchResult {
    for (const actionTaken of store.getActions()) {
        if (actionTaken.type === actionType) {
            return {
                message: () => `Result contained action type ${actionType}`,
                pass: true
            };
        }
    }

    return {
        message: () => `Could not find action type ${actionType}`,
        pass: false
    };
}

expect.extend({
    toPointTo,
    toHaveExitThatPointsTo,
    toHaveInboundFrom,
    toHaveExitWithDestination,
    toHaveInboundConnections,
    toHavePayload,
    toHaveReduxAction
});
