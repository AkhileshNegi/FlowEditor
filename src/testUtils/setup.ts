import { configure } from 'enzyme';
import * as Adapter from 'enzyme-adapter-react-16';
import * as $ from 'jquery';

// Configure Enzyme adapter
configure({ adapter: new Adapter() });

// RAF shim
(global as any).requestAnimationFrame = callback => {
    setTimeout(callback, 0);
};

// Make jest aware of our jQuery dep
(window as any).$ = $;
