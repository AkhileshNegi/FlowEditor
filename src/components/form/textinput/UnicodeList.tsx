import * as React from 'react';
import { pluralize, UnicodeCharMap } from '~/components/form/textinput/helpers';
import * as styles from '~/components/form/textinput/UnicodeList.scss';
import { renderIf } from '~/utils';

export interface UnicodeListProps {
    unicodeChars: UnicodeCharMap;
}

export const utfWarning = (num: number) =>
    ` Note that your message may require more segments to send because it contains the following ${pluralize(
        num,
        'character'
    )}:`;

export const unicodeListContainerSpecId = 'unicode-list-container';
export const unicodeListSpecId = 'unicode-list';

const UnicodeList: React.SFC<UnicodeListProps> = ({ unicodeChars }) => {
    const chars = Object.keys(unicodeChars);
    return renderIf(chars.length > 0)(
        <span data-spec={unicodeListContainerSpecId}>
            {utfWarning(chars.length)}
            <div className={styles.unicodeList} data-spec={unicodeListSpecId}>
                {chars.map((char, idx) => (
                    <span key={char} className={styles.unicodeChar}>
                        {char}
                    </span>
                ))}
            </div>
        </span>
    );
};

export default UnicodeList;
