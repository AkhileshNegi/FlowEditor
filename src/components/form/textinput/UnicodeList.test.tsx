import UnicodeList, { UnicodeListProps, utfWarning } from 'components/form/textinput/UnicodeList';
import React from 'react';
import { render } from 'react-testing-library';
import { composeComponentTestUtils } from 'testUtils';

const baseProps: UnicodeListProps = {
  unicodeChars: {
    '😎': true,
    '👍': true
  }
};

const { setup } = composeComponentTestUtils<UnicodeListProps>(UnicodeList, baseProps);

describe('UnicodeList', () => {
  describe('helpers', () => {
    describe('utfWarning', () => {
      it('should pluralize "character" if passed a number other than 1', () => {
        expect(utfWarning(2).indexOf('characters') > -1).toBeTruthy();
      });

      it("shouldn't pluralize 'character' if passed the number 1", () => {
        expect(utfWarning(1).indexOf('character') > -1).toBeTruthy();
      });
    });
  });

  describe('render', () => {
    it('should render self, children with base props', () => {
      const { getByTestId } = render(<UnicodeList {...baseProps} />);
      expect(getByTestId('unicode-list-container')).toMatchSnapshot();
    });
  });
});
