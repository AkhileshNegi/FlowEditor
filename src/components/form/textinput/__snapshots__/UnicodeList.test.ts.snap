// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UnicodeList render should render self, children with base props 1`] = `
<span
  data-spec="unicode-list-container"
>
   Note that your message may require more segments to send because it contains the following characters:
  <div
    className="unicodeList"
    data-spec="unicode-list"
  >
    <span
      className="unicodeChar"
      key="😎"
    >
      😎
    </span>
    <span
      className="unicodeChar"
      key="👍"
    >
      👍
    </span>
  </div>
</span>
`;
